### PR TITLE
feat: add S2Vec embedding model

### PR DESCRIPTION
**Pull request by hubkrieb**
S2Vec is a recent geospatial embedding model proposed by Google Research that relies on S2 cells and the Masked Autoencoder architecture.

I have implemented it into the library and added a small example following the one existing for other models (using larger S2 resolution and a larger tag set than in the original paper).

### CODE REVIEW

**Comment by RaczeQ** [comment]
Hello @hubkrieb, thank you for this contribution!

We will try to make a proper review of this code, but on a first glance, it seems like you didn't commit the new required libraries (`from timm.models.vision_transformer import Block`) and there are no tests for the model.

Do you need help with adding them to the pyproject and writing the tests for the new module?

**Comment by hubkrieb** [comment]
Hi @RaczeQ I just added the tests for the S2Vec embedder. One thing to notice is that I couldn't make the ViT Block from `timm` fully deterministic so I didn't manage to have a threshold as low as for other embedders in the `test_embedder_save_load`.

**Comment by nside** [comment]
Would be nice to have a notebook with the downstream eval per the paper to see how it compares

**Comment by nside** [comment]
got it. why not merge it?

**Comment by hubkrieb** [comment]
I'd be happy to contribute to a benchmark but I agree that it might be a bit too much for this PR.

About a simple baseline, I don't know if you had anything in mind @RaczeQ  apart from the simple examples with embeddings visualization

**Comment by RaczeQ** [comment]
This kind of example is enough, so don't worry @hubkrieb 
The benchmark PR is mostly done, just have to do some final code review. You can look into it here: #521

**Comment by RaczeQ** [comment]
Hi @hubkrieb, sorry for the very long delay, but I have finally found time to dig into the implementation.

It's great and basically follows what I understood from the paper about their logic.

I have seen small discrepancies between the paper and this implementation:
- Default number of attention heads in the paper is 8, in the model is 8, but the embedder has a default of 2.
- Their implementation includes normalisation with mean and std before training
- They also have a dropout in the model.

I started testing the model with different parameters.
One thing that stood out to me was the harsh lines between `images`. I checked if the added normalisation changes the output. I have attached screenshots below.

Without normalisation:
image

With normalisation:
image

The change between images is now much more gradual.

And the other city example (Warsaw, Poland):

Without normalisation:
image

With normalisation:
image


---

I have tried changing the `img_resolution` from 12 to 13, while keeping the `patch_resolution` at 16 in the example notebook. Since the area in the example is so small, I wanted to experiment with a higher number of images, while sacrificing the number of patches (reduced from 256 to 64 per image).

No normalisation:
image

With normalisation before training:
image

The embeddings in the centre are more nuanced in the normalised version and not squashed like in the current version. But this change reintroduced harsh boundaries between images.

---

In summary, thank you for this contribution. It seems like everything is working, but it's hard to evaluate if the visible harsh differences between images are by design, or maybe there is a sneaky mistake in the implementation, although it seems like normalisation will help with that.

I have pushed some changes:
- Added dropout
- Added normalisation
- Moved `timm` import inside class `__init__` function
- Fixed saving logic with additional parameters in the `model_config`
- Fixed tests by recalculating embeddings and lowering the tolerance.

**Comment by RaczeQ** [comment]
I have looked into determinism in tests and was able to solve it by adding weights init to each nn.Linear layer in the model (there were 3) and its working across operating systems.

**Review by RaczeQ — APPROVED** [review]

**Comment by RaczeQ** [comment]
Thank you @hubkrieb for this implementation, and sorry again that the review process took so long.

I had to consult someone from our team to also look at the code, and now we are able to merge it.
